### ISSUE TITLE
Closing socket can cause connection reset.

### ISSUE THREAD

**Issue by iamthechad** [iamthechad/sslinfo#5]
Against some servers, many `ECONNRESET` instances are seen.

The socket connection code looks like this: 

```
var socket = tls.connect(fullOptions, function() {
    deferred.resolve({protocol: options.secureProtocol, protocolCommonName: options.protocolCommonName, cipher: options.ciphers, enabled: true});
    socket.end();
});
```

Calling `socket.end()` sends a `FIN` to the server. Some servers respond with `FIN ACK`, and it appears that a large number of servers respond by simply resetting the connection. This reset is treated as an error on the client side, incorrectly triggering the error handling and probably resulting in incorrect results.

It seems that there are two solutions:
1. Track the successful connection status to see if the `ECONNRESET` is just a server terminating the connection
2. Call `socket.destroy()`, which will reset the connection from the client end.

**Comment by chrisDeFouRire** [comment]
but then a promise can be fulfilled only once… and socket.end is called after deferred.resolve.

but maybe it could help to remove the listener from the socket before calling end() ? (with socket.removeAllListeners())

**Comment by chrisDeFouRire** [comment]
In addition to sslinfo, I'm also using tls.connect in my project and was experiencing occasional ECONNRESET... I seem to have fixed it now, with the following pattern :

Sorry, it's coffeescript...

```
    new Q (resolve, reject)->

        socket = tls.connect {host, port, rejectUnauthorized: true, servername: host}, (error)->
            if error?
                reject error
            else
                socket.removeAllListeners()
                socket.end()
                socket.destroy()
                resolve null

        socket.once 'error', (error)->
# some handling here
            reject error
            socket.destroy()

```

I guess the socket.removeAllListeners() makes a difference...
It might also fix the memory leak someone has mentioned, by not leaking listeners.
Also notice the socket.once... which is more a matter of logic: promises can be resolved or rejected only once anyway...

**Comment by chrisDeFouRire** [comment]
Actually `.once` causes more problems...

I wonder if some kind of DDoS protection could be at play. 
When trying `bitcoinwisdom.com` I get ECONNREFUSED errors... sslinfo does establish a lot of short lived connections... just like a DDoS would do.

**Comment by iamthechad** [comment]
I switched up the order of socket closing and resolving, plus added a check for connection refused. In my testing, it worked against `microsoft.com`, `bitcoinwisdom.com` and `MDM.VILLAGECARE.ORG`.

**Comment by chrisDeFouRire** [comment]
I think some "errors" may be caused by anti DoS rules (limit connections to n per sec from the same IP)... Another class of errors I've seen: ENOTFOUND, or DNS errors ! I guess sslinfo is not caching DNS resolution.
I've looked into sslscan, and it appears it has an option to delay connections, maybe to mitigate DoS issues... I'll tell you how it goes with this new version ! Thanks for your help !

**Comment by chrisDeFouRire** [comment]
seems to work fine ! I'll tell you if I find more problems! Thanks
